File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)

        prefix = [1] * n
        suffix = [1] * n

        prefix[0] = nums[0]
        suffix[n - 1] = nums[n - 1]

        for i in range(1, n):
            prefix[i] = prefix[i - 1] * nums[i]

        for i in range(n - 2, -1, -1):
            suffix[i] = suffix[i + 1] * nums[i]

        result = [0] * n

        result[0] = suffix[1]
        result[n - 1] = prefix[n - 2]

        for i in range(1, n - 1):
            result[i] = prefix[i - 1] * suffix[i + 1]

        return result
        

File: Data Structures & Algorithms/three-integer-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        //  -4 -1 -1 0 1 2
        sort(nums.begin(), nums.end(), cmp);
        vector<vector<int>> result;
        int len = nums.size();
        for (int i = 0; i < len; i++) {
            
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            
            int l = i+1, r = len-1;
            while (l < r) {
            int sum = nums[l] + nums[i] + nums[r];
                if (sum == 0) {
                    vector<int> li = {nums[l] , nums[i] , nums[r]};
                    result.push_back(li);
                    l++;
                    r--;

                    while (l < r && nums[l] == nums[l-1])
                        l++;
                    while (l < r && nums[r] == nums[r+1])
                        r--;
                    
                } else if (sum > 0) {
                    r--;
                } else {
                    l++;
                }
            }
        }
        return result;    
    }

    static bool cmp(int& a, int& b) {
        return a < b;
    }
};
